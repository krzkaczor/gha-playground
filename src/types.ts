import { exec } from '@actions/exec'

export type Exec = typeof exec
