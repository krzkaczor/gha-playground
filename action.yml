name: 'Hello World'
description: 'Greet someone and record the time'
inputs:
  github_token:
    required: true
runs:
  using: 'node12'
  main: 'dist/index.js'
