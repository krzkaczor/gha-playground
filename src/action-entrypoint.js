const { join } = require('path')
require('ts-node').register({ project: join(__dirname, '../tsconfig.prod.json') })

const { action } = require('index.ts')

const core = require('@actions/core')
action().catch((e) => {
  console.log('Error happened:', e)
  // in case of error mark action as failed
  core.setFailed(e.message)
})
