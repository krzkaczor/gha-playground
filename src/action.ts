import { copySync } from 'fs-extra'
import { join } from 'path'
import { dirSync as tmp } from 'tmp'

import { Exec } from './exec'
import { newPristineBranch, setup } from './git/utils'

interface ActionCtx {
  exec: Exec
  env: NodeJS.ProcessEnv
  cwd: string
}

interface Options {
  branchName: string
  files: string[]
}

export async function action(ctx: ActionCtx, options: Options) {
  const tmpDir = tmp().name
  console.log('Tmp dir: ', tmpDir)

  // copy deploy files away
  for (const file of options.files) {
    const fullOutputPath = join(tmpDir, file)

    // ensureDirSync(dirname(fullOutputPath))
    copySync(join(ctx.cwd, file), fullOutputPath)
  }

  await setup(ctx.exec, ctx.env)
  await newPristineBranch(ctx.exec, options.branchName)

  // copy files back
  for (const file of options.files) {
    await ctx.exec(`git add --force ${file}`)
  }

  await ctx.exec(`git commit -m "Automated release" -a`)
  await ctx.exec(`git push origin-authorized`)
}
