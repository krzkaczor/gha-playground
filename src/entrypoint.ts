import * as core from '@actions/core'
import * as github from '@actions/github'
import { assert } from 'ts-essentials'

import { addContributions } from './all-contributors/cli/addContributors'
import { findConfig } from './all-contributors/cli/findConfig'
import { generateContributorsListIntoMarkdown } from './all-contributors/cli/generate'
import { parseComment } from './all-contributors/parseComment'
import { pushAll } from './git/push-all'
import { addReaction } from './github/addReaction'
import { checkPermissions } from './github/checkPermissions'
import { getCommentBody, getCommentId } from './github/getters'
import { getOctokit } from './github/octokit'

export async function action() {
  const ctx = github.context
  const octokit = getOctokit()
  const cwd = process.cwd()

  const comment = getCommentBody(ctx)
  if (!comment) {
    throw new Error("Comment body couldn't be found. Did you setup action to run on `issue_comment` event?")
  }

  const actorPermissions = await checkPermissions(octokit, github.context, github.context.actor)
  if (actorPermissions !== 'write' && actorPermissions !== 'admin') {
    throw new Error(`${github.context.actor} doesn't have access to run this action`)
  }

  const configPath = findConfig(cwd)
  if (!configPath) {
    throw new Error("Can't find all contributors config file!")
  }

  const contributions = parseComment(comment)
  console.log(JSON.stringify(contributions))
  await addContributions(configPath, contributions)
  await generateContributorsListIntoMarkdown({ configPath, cwd })

  await pushAll()

  const commentId = getCommentId(ctx)
  assert(
    commentId !== undefined,
    "Comment body couldn't be found. Did you setup action to run on `issue_comment` event?",
  )
  await addReaction(octokit, ctx, commentId)
}

action().catch((e) => {
  console.log('Error happened:', e)
  // in case of error mark action as failed
  core.setFailed(e.message)
})
