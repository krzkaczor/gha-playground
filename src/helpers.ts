import { context } from '@actions/github'
import { assert } from 'ts-essentials'

type Context = typeof context

export function getContextBody(ctx: Context): string | undefined {
  const body = ctx.payload.comment?.body

  assert(body !== undefined, "Comment body couldn't be found. Did you setup action to run on `issue_comment` event?")

  return body
}
