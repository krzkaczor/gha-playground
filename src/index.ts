import * as core from '@actions/core'

export function action() {
  // `who-to-greet` input defined in action metadata file
  const nameToGreet = core.getInput('who-to-greet')
  console.log(`Hello ${nameToGreet}!`)
  const time = new Date().toTimeString()
  core.setOutput('time', time)
}
