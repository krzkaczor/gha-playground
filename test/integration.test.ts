import { ensureDirSync, writeFileSync } from 'fs-extra'
import { dirname, join } from 'path'
import simpleGit from 'simple-git/promise'
import { dirSync as tmp } from 'tmp'
import { expect } from 'earljs'

import { action } from '../src/action'
import { Exec, getExec } from '../src/exec'

describe('integration', () => {
  it('works', async () => {
    const workspace = tmp().name
    console.log('Workspace: ', workspace)
    const git = simpleGit(workspace)
    const workspaceFiles = {
      'action.yml': `name: 'action for tests'`,
      'dist/index.js': `console.log('test!')`,
      '.gitignore': 'dist/index.js',
    }
    writeFiles(workspaceFiles, workspace)
    await git.init()
    const realExec = getExec(workspace)
    await realExec('git add .gitignore action.yml')
    await realExec('git commit -m init')

    const filteringExec: Exec = (...args: any[]) => {
      // ignore git pushes
      if (args[0].startsWith('git push')) {
        return // do nothing because push will fail anyway
      }
      return (realExec as any)(...args)
    }

    await action(
      { cwd: workspace, env: {}, exec: filteringExec },
      { branchName: 'action', files: ['action.yml', 'dist/index.js'] },
    )

    const exactOutput = await realExec('git diff-tree --no-commit-id --name-status -r HEAD')
    expect(exactOutput).toMatchSnapshot()
  })
})

function writeFiles(workspaceFiles: Record<string, string>, path: string) {
  for (const [filePath, contents] of Object.entries(workspaceFiles)) {
    const fullPath = join(path, filePath)
    ensureDirSync(dirname(fullPath))
    writeFileSync(fullPath, contents)
  }
}
