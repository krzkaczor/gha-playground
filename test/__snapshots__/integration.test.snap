// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration works 1`] = `
"A	action.yml
A	dist/index.js
"
`;
